ape loader fails with busybox dd

in a dockerfile like so:
```Dockerfile
FROM alpine:3

RUN apk add git make bash zip
RUN git clone --depth=1 https://github.com/jart/cosmopolitan.git /build/cosmopolitan
RUN make -C /build/cosmopolitan -j -l4 -O O=o//tool/net/redbean.com
```
we see:
```
o//third_party/quickjs/qjsc.com.tmp.68677: exec: line 13: /build/cosmopolitan/o/tmp/ape-loader: not found

`make MODE= -j16 o//third_party/quickjs/qjscalc.c` exited with 127:
o//third_party/quickjs/qjsc.com -fbignum -o o//third_party/quickjs/qjscalc.c -c third_party/quickjs/qjscalc.js
consumed 634µs wall time
ballooned to 1,088kb in size
needed 595us cpu (0% kernel)
caused 133 page faults (100% memcpy)
7 context switch (100% consensual)
performed 160 read and 0 write i/o operations

make: *** [third_party/quickjs/quickjs.mk:134: o//third_party/quickjs/qjscalc.c] Error 127
make: Entering directory '/build/cosmopolitan'
make: Leaving directory '/build/cosmopolitan'
The command '/bin/sh -c make -C /build/cosmopolitan -j -l4 -O O=o//tool/net/redbean.com' returned a non-zero code: 2
```

poking around the build dir by hand we can uncover:
```
o//test/libc/release/smoke-nms.com.tmp.70227: exec: line 13: /build/cosmopolitan/o/tmp/ape-loader: not found

`make MODE= -j16 o//test/libc/release/smoke-nms.com.runs` exited with 127:
o//test/libc/release/smoke-nms.com
consumed 634µs wall time
ballooned to 1,100kb in size
needed 546us cpu (0% kernel)
caused 132 page faults (100% memcpy)
7 context switch (100% consensual)
performed 160 read and 0 write i/o operations

make: *** [build/rules.mk:76: o//test/libc/release/smoke-nms.com.runs] Error 127
make: Leaving directory '/build/cosmopolitan'
```

stracing the build with `strace -f -o /tmp/strace.txt make -C /build/cosmopolitan/ MODE= -j16 o//test/libc/release/smoke-nms.com.runs` shows:
```
/ # grep dd /tmp/strace.txt | tail -n 10
70257 set_tid_address(0x7ffb9a458f90)   = 70257
70257 stat("/usr/local/sbin/dd", 0x7ffc5f877c40) = -1 ENOENT (No such file or directory)
70257 stat("/usr/local/bin/dd", 0x7ffc5f877c40) = -1 ENOENT (No such file or directory)
70257 stat("/usr/sbin/dd", 0x7ffc5f877c40) = -1 ENOENT (No such file or directory)
70257 stat("/usr/bin/dd", 0x7ffc5f877c40) = -1 ENOENT (No such file or directory)
70257 stat("/sbin/dd", 0x7ffc5f877c40)  = -1 ENOENT (No such file or directory)
70257 stat("/bin/dd", {st_mode=S_IFREG|0755, st_size=824984, ...}) = 0
70259 execve("/bin/dd", ["dd", "if=o//test/libc/release/smoke-nm"..., "of=/build/cosmopolitan/o/tmp/ape"..., "skip=    3200", "count=      28", "bs=64"], 0x7ffb9a3be2f8 /* 12 vars */) = 0
70259 set_tid_address(0x7f58fed0bf90)   = 70259
70259 write(2, "dd: invalid number '    3200'\n", 30) = 30
```

if i `apk add coreutils`, it succeeds.

i looked at the code in ape/ape.S, but i do not understand where `.shstub` assembly directive comes from, but it seems that busybox dd doesnt like spaces in number arguments, like
```
/ # dd 'skip=       1'
dd: invalid number '       1'
```

can we fix the script in such a way that it is compatible with busybox dd?

--- a/ape/ape.S
+++ b/ape/ape.S
@@ -569,11 +569,11 @@ apesh:	.ascii	"'\n#'\"\n"			# sixth edition shebang
 	.ascii	  "t=\"${TMPDIR:-${HOME:-.}}/.ape\"\n"
 	.ascii	  "[ -x \"$t\" ] || {\n"
 	.ascii	    "mkdir -p \"${t%/*}\" &&\n"
-	.ascii	    "dd if=\"$o\" of=\"$t.$$\" skip=\""
+	.ascii	    "dd if=\"$o\" of=\"$t.$$\" skip="
 	.shstub	      ape_loader_dd_skip,2
-	.ascii	      "\" count=\""
+	.ascii	      " count="
 	.shstub	      ape_loader_dd_count,2
-	.ascii	      "\" bs=64 2>/dev/null\n"
+	.ascii	      " bs=64 2>/dev/null\n"
 #if SupportsXnu()
 	.ascii	    "[ -d /Applications ] && "
 	.ascii	      "dd if=\"$t.$$\""
@@ -637,11 +637,11 @@ apesh:	.ascii	"'\n#'\"\n"			# sixth edition shebang
 	.ascii	  "dd if=\"$o\""
 	.ascii	    " of=\"$o\""
 	.ascii	    " bs=8"
-	.ascii	    " skip=\""
+	.ascii	    " skip="
 	.shstub	    ape_macho_dd_skip,2
-	.ascii	    "\" count=\""
+	.ascii	    " count="
 	.shstub	    ape_macho_dd_count,2
-	.ascii	    "\" conv=notrunc 2>/dev/null\n"
+	.ascii	    " conv=notrunc 2>/dev/null\n"
 #endif /* XNU */
 	.ascii	"[ x\"$1\" = x--assimilate ] && exit 0\n"
 #ifndef APE_NO_MODIFY_SELF

--- a/ape/macros.internal.h
+++ b/ape/macros.internal.h
@@ -221,9 +221,12 @@
 /**
  * Binary coded decimal support.
  *
- * <p>This allows linker scripts to generate dd commands. Leading spaces
- * need to be inserted so Mac doesn't consider them octal; therefore,
- * parameters must be quoted; and eight digits should be good enough.
+ * <p>This allows linker scripts to generate dd commands, e.g. ape.lds.
+ * There are a few ways to pad each number to the necessary 8 bytes.
+ * Spaces cannot be prepended because busybox refuses to parse them.
+ * Zeros cannot be prepended because Mac will take numbers as octal.
+ * That leaves appending spaces. The user's shell ought to treat any
+ * unquoted run of spaces as if there was only one, so this is safe.
  */
 #define SHSTUB2(SYM, X)             \
   HIDDEN(SYM##_bcs0 = BCD_LEFT(X)); \

--- a/ape/macros.internal.h
+++ b/ape/macros.internal.h
@@ -225,40 +225,39 @@
  * need to be inserted so Mac doesn't consider them octal; therefore,
  * parameters must be quoted; and eight digits should be good enough.
  */
-#define SHSTUB2(SYM, X)           \
-  HIDDEN(SYM##_bcs0 = BCD10K(X)); \
-  HIDDEN(SYM##_bcs1 = BCD(X))
-#define BCD(X)                                                                 \
-  ((X) == 0                                                                    \
-       ? 0x20202030                                                            \
-       : (X) < 10 ? 0x30202020 + (((X) % 10) << 24)                            \
-                  : (X) < 100 ? 0x30302020 + (((X) % 10) << 24) +              \
-                                    (((X) / 10 % 10) << 16)                    \
-                              : (X) < 1000 ? 0x30303020 + (((X) % 10) << 24) + \
-                                                 (((X) / 10 % 10) << 16) +     \
-                                                 (((X) / 100 % 10) << 8)       \
-                                           : 0x30303030 + (((X) % 10) << 24) + \
-                                                 (((X) / 10 % 10) << 16) +     \
-                                                 (((X) / 100 % 10) << 8) +     \
-                                                 (((X) / 1000 % 10) << 0))
-#define BCD10K(X)                                                          \
-  ((X) < 10000                                                             \
-       ? 0x20202020                                                        \
-       : (X) < 100000                                                      \
-             ? 0x30202020 + (((X) / 10000 % 10) << 24)                     \
-             : (X) < 1000000                                               \
-                   ? 0x30302020 + (((X) / 10000 % 10) << 24) +             \
-                         (((X) / 100000 % 10) << 16)                       \
-                   : (X) < 10000000                                        \
-                         ? 0x30303020 + (((X) / 10000 % 10) << 24) +       \
-                               (((X) / 100000 % 10) << 16) +               \
-                               (((X) / 1000000 % 10) << 8)                 \
-                         : (X) < 100000000                                 \
-                               ? 0x30303030 + (((X) / 10000 % 10) << 24) + \
-                                     (((X) / 100000 % 10) << 16) +         \
-                                     (((X) / 1000000 % 10) << 8) +         \
-                                     (((X) / 10000000 % 10) << 0)          \
-                               : 0xffffffffffffffff)
+#define SHSTUB2(SYM, X)             \
+  HIDDEN(SYM##_bcs0 = BCD_LEFT(X)); \
+  HIDDEN(SYM##_bcs1 = BCD_RIGHT(X))
+#define BCD_HELPER(X)                          \
+  (((X)) < 10    ? 0x20202030 +                \
+                   (X) % 10                    \
+   : (X) < 100   ? 0x20203030 +                \
+                   ((X) / 10) % 10 +           \
+                   (X) % 10 * 0x100            \
+   : (X) < 1000  ? 0x20303030 +                \
+                   ((X) / 100) % 10 +          \
+                   ((X) / 10) % 10 * 0x100 +   \
+                   (X) % 10 * 0x10000          \
+   : (X) < 10000 ? 0x30303030 +                \
+                   ((X) / 1000) % 10 +         \
+                   ((X) / 100) % 10 * 0x100 +  \
+                   ((X) / 10) % 10 * 0x10000 + \
+                   (X) % 10 * 0x1000000        \
+   : 0xffffffffffffffff)
+#define BCD_LEFT(X)                            \
+  (((X)) < 10000     ? BCD_HELPER(X)           \
+   : (X) < 100000    ? BCD_HELPER((X) / 10)    \
+   : (X) < 1000000   ? BCD_HELPER((X) / 100)   \
+   : (X) < 10000000  ? BCD_HELPER((X) / 1000)  \
+   : (X) < 100000000 ? BCD_HELPER((X) / 10000) \
+   : 0xffffffffffffffff)
+#define BCD_RIGHT(X)                           \
+  (((X)) < 10000     ? 0x20202020              \
+   : (X) < 100000    ? BCD_HELPER((X) % 10)    \
+   : (X) < 1000000   ? BCD_HELPER((X) % 100)   \
+   : (X) < 10000000  ? BCD_HELPER((X) % 1000)  \
+   : (X) < 100000000 ? BCD_HELPER((X) % 10000) \
+   : 0xffffffffffffffff)
 
 #endif /* __ASSEMBLER__ */
 #endif /* APE_MACROS_H_ */

--- a/ape/macros.internal.h
+++ b/ape/macros.internal.h
@@ -231,36 +231,31 @@
 #define SHSTUB2(SYM, X)             \
   HIDDEN(SYM##_bcs0 = BCD_LEFT(X)); \
   HIDDEN(SYM##_bcs1 = BCD_RIGHT(X))
-#define BCD_HELPER(X)                          \
-  (((X)) < 10    ? 0x20202030 +                \
-                   (X) % 10                    \
-   : (X) < 100   ? 0x20203030 +                \
-                   ((X) / 10) % 10 +           \
-                   (X) % 10 * 0x100            \
-   : (X) < 1000  ? 0x20303030 +                \
-                   ((X) / 100) % 10 +          \
-                   ((X) / 10) % 10 * 0x100 +   \
-                   (X) % 10 * 0x10000          \
-   : (X) < 10000 ? 0x30303030 +                \
-                   ((X) / 1000) % 10 +         \
-                   ((X) / 100) % 10 * 0x100 +  \
-                   ((X) / 10) % 10 * 0x10000 + \
-                   (X) % 10 * 0x1000000        \
-   : 0xffffffffffffffff)
-#define BCD_LEFT(X)                            \
-  (((X)) < 10000     ? BCD_HELPER(X)           \
-   : (X) < 100000    ? BCD_HELPER((X) / 10)    \
-   : (X) < 1000000   ? BCD_HELPER((X) / 100)   \
-   : (X) < 10000000  ? BCD_HELPER((X) / 1000)  \
-   : (X) < 100000000 ? BCD_HELPER((X) / 10000) \
-   : 0xffffffffffffffff)
-#define BCD_RIGHT(X)                           \
-  (((X)) < 10000     ? 0x20202020              \
-   : (X) < 100000    ? BCD_HELPER((X) % 10)    \
-   : (X) < 1000000   ? BCD_HELPER((X) % 100)   \
-   : (X) < 10000000  ? BCD_HELPER((X) % 1000)  \
-   : (X) < 100000000 ? BCD_HELPER((X) % 10000) \
-   : 0xffffffffffffffff)
+#define BCD_SMEAR(X) ((X) + (X) * 10000)
+#define BCD_LEFT(X)                                      \
+  (((X)) < 10000     ? BCD_RIGHT(BCD_SMEAR(X)) | 0x10    \
+   : (X) < 100000    ? BCD_RIGHT(BCD_SMEAR((X) / 10))    \
+   : (X) < 1000000   ? BCD_RIGHT(BCD_SMEAR((X) / 100))   \
+   : (X) < 10000000  ? BCD_RIGHT(BCD_SMEAR((X) / 1000))  \
+   : (X) < 100000000 ? BCD_RIGHT(BCD_SMEAR((X) / 10000)) \
+                     : 0xffffffffffffffff)
+#define BCD_RIGHT(X) \
+  (((X)) < 10000     ? 0x20202020                  \
+   : (X) < 100000    ? 0x20202030 +                \
+                       (X) % 10                    \
+   : (X) < 1000000   ? 0x20203030 +                \
+                       ((X) / 10) % 10 +           \
+                       (X) % 10 * 0x100            \
+   : (X) < 10000000  ? 0x20303030 +                \
+                       ((X) / 100) % 10 +          \
+                       ((X) / 10) % 10 * 0x100 +   \
+                       (X) % 10 * 0x10000          \
+   : (X) < 100000000 ? 0x30303030 +                \
+                       ((X) / 1000) % 10 +         \
+                       ((X) / 100) % 10 * 0x100 +  \
+                       ((X) / 10) % 10 * 0x10000 + \
+                       (X) % 10 * 0x1000000        \
+                     : 0xffffffffffffffff)
 
 #endif /* __ASSEMBLER__ */
 #endif /* APE_MACROS_H_ */

